### ISSUE TITLE
Calculate F1 score in custom training loop

### ISSUE THREAD

**Issue by Eason51** [keras-team/keras#17214]
I wonder how I can implement a F1score class that inherits from tf.keras.metrics.Metric so that I can use this class to calculate F1score in a custom training loop.

In the following code segment, I have a bunch of metrics, e.g. binaryaccuracy, precision, recall. I calculate these metrics by calling the "update_state()", "result()", "reset_state()" methods of each of the metric class in the appropriate location inside of the custom training loop. My objective is that I want do exactly the same thing to calculate the f1score, but I do not know how to implement by extending the Metric class

```
		# Prepare the metrics.
		train_acc_metric = keras.metrics.BinaryAccuracy()
		val_acc_metric = keras.metrics.BinaryAccuracy()
		val_auc_metric = keras.metrics.AUC(from_logits=True)
		val_precision_metric = keras.metrics.Precision()
		val_recall_metric = keras.metrics.Recall()
		# TODO: Find a method to calculate f1scores
		# val_f1_metric = ???

		val_epoch_loss_avg = keras.metrics.Mean()


		@tf.function
		def run_train_step(x_batch_train, y_batch_train):
			with tf.GradientTape() as tape:
				logits = model(x_batch_train, training=True)
				loss_value = loss_fn(y_batch_train, logits)
			grads = tape.gradient(loss_value, model.trainable_weights)
			optimizer.apply_gradients(zip(grads, model.trainable_weights))

			# Update training metric.
			train_acc_metric.update_state(y_batch_train, logits)

			return loss_value

		# Assign the model to the callbacks.
		for callback in callbacks:
			callback.model = model

		@tf.function
		def run_val_step(x_batch_val, y_batch_val):
			val_logits = model(x_batch_val, training=False)
			loss_value = loss_fn(y_batch_val, val_logits)
			# Update val metrics
			val_acc_metric.update_state(y_batch_val, val_logits)
			val_auc_metric.update_state(y_batch_val, val_logits)
			val_precision_metric.update_state(y_batch_val, val_logits)
			val_recall_metric.update_state(y_batch_val, val_logits)
			# val_f1_metric.update_state(y_batch_val, val_logits)
			val_epoch_loss_avg.update_state(loss_value)

			return loss_value


		# Record the best validation loss value
		best_epoch_loss = float("inf")

		epochs = 10
		for epoch in range(epochs):
			# Iterate over the batches of the dataset.
			for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
				loss_value = run_train_step(x_batch_train, y_batch_train)

				# Log every 200 batches.
				if step % 200 == 0:
					print(
						"Training loss (for one batch) at step %d: %.4f"
						% (step, float(loss_value))
					)
					print("Seen so far: %d samples" % ((step + 1) * batch_size))

			# Display metrics at the end of each epoch.
			train_acc = train_acc_metric.result()
			print("Training acc over epoch: %.4f" % (float(train_acc),))

			# Reset training metrics at the end of each epoch
			train_acc_metric.reset_states()


			# Run a validation loop at the end of each epoch.
			for x_batch_val, y_batch_val in val_dataset:
				val_loss_value = run_val_step(x_batch_val, y_batch_val)

			val_acc = val_acc_metric.result()
			val_auc = val_auc_metric.result()
			val_precision = val_precision_metric.result()
			val_recall = val_recall_metric.result()
			# val_f1 = val_f1_metric.result()
			val_loss = float(val_epoch_loss_avg.result().numpy())
			for callback in callbacks:
				# callback.on_epoch_end(epoch, logs={"val_loss": float(format(float(val_loss), ".4f"))})
				callback.on_epoch_end(epoch, logs={"epoch_loss": val_loss})
			val_acc_metric.reset_states()
			val_auc_metric.reset_states()
			val_precision_metric.reset_states()
			val_recall_metric.reset_states()
			# val_f1_metric.reset_states()
			val_epoch_loss_avg.reset_states()
			print("Validation acc: %.4f" % (float(val_acc),))
			print("Validation auc: %.4f" % (float(val_auc),))
			print("Validation precision: %.4f" % (float(val_precision),))
			print("Validation recall: %.4f" % (float(val_recall),))
			# print("Validation f1: %.4f" % (float(val_f1),))
			print("Validation loss: %.4f" % (float(val_loss),))

```

**Comment by bhack** [comment]
Check https://github.com/keras-team/keras-cv/pull/423
